### PR TITLE
Add bevy_picking sprite backend

### PR DESCRIPTION
**Pull request by jnhyatt**
# Objective

Add `bevy_picking` sprite backend as part of the `bevy_mod_picking` upstreamening (#12365).

## Solution

More or less a copy/paste from `bevy_mod_picking`, with the changes [here](https://github.com/aevyrie/bevy_mod_picking/pull/354). I'm putting that link here since those changes haven't yet made it through review, so should probably be reviewed on their own.

## Testing

I couldn't find any sprite-backend-specific tests in `bevy_mod_picking` and unfortunately I'm not familiar enough with Bevy's testing patterns to write tests for code that relies on windowing and input. I'm willing to break the pointer hit system into testable blocks and add some more modular tests if that's deemed important enough to block, otherwise I can open an issue for adding tests as follow-up.

## Follow-up work

- More docs/tests
- Ignore pick events on transparent sprite pixels with potential opt-out

### CODE REVIEW

**Comment by aevyrie** [comment]
This doesn't test against alpha, which might be surprising in a first-party implementation. Would be nice to document on the plugin.

Additionally, it would be courteous to co-author the downstream contributors in a commit here.

File: crates/bevy_sprite/src/picking_backend.rs [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+use bevy_asset::prelude::*;
+use bevy_ecs::prelude::*;
+use bevy_math::prelude::*;
+use bevy_picking::backend::prelude::*;
+use bevy_render::prelude::*;
+use bevy_transform::prelude::*;
+use bevy_window::PrimaryWindow;
+
+#[derive(Clone)]
+pub struct SpriteBackend;
```
Comment (aevyrie) [inline crates/bevy_sprite/src/picking_backend.rs:16]: Could be helpful to add picking to the name, if I saw `SpriteBackend` I wouldn't think it was for picking. Not sure if there are any established conventions for this yet.

**Review by aevyrie — APPROVED** [review]

**Comment by jnhyatt** [comment]
> This doesn't test against alpha, which might be surprising in a first-party implementation. Would be nice to document on the plugin.

Not sure what you mean by "test against alpha". Could you elaborate?
> Additionally, it would be courteous to co-author the downstream contributors in a commit here.

Absolutely! I thought I'd opened this in draft and I hadn't planned on it getting merged without crediting you. Since you're the original author, is there anyone else you'd like listed as co-author? I also have to say I've never added a co-author, so I might mess it up a couple times, haha

**Comment by aevyrie** [comment]
> Not sure what you mean by "test against alpha". Could you elaborate? EDIT - By that, do you mean whether or not the sprite is visible due to alpha/invisible texels? I'll add docs for that.

Exactly. I think most users would expect fully transparent pixels to not be hittable. Might be something where the backend/component specifies a threshold, so any alpha below that value will be ignored. Doesn't need to be added right now, but would be nice to mention in the plugin docs.

> is there anyone else you'd like listed as co-author? I also have to say I've never added a co-author, so I might mess it up a couple times, haha

There are a bunch of people who have contributed to this backend. I'd take a look at the commit history here - I'm not sure if there is an easier way to do this: https://github.com/aevyrie/bevy_mod_picking/commits/main/backends/bevy_picking_sprite/src/lib.rs

No worries. I think git makes it pretty simple, you just need to add the names/emails separated by newlines in a commit.

File: crates/bevy_sprite/src/picking_backend.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                        world_to_sprite.transform_point3(cursor_ray_world.origin);
+                    let cursor_end_sprite = world_to_sprite.transform_point3(cursor_ray_end);
+
+                    fn inverse_lerp(a: f32, b: f32, x: f32) -> Option<f32> {
+                        if (b - a).abs() < f32::EPSILON {
+                            None
+                        } else {
+                            Some((x - a) / (b - a))
+                        }
+                    }
```
Comment (kristoff3r) [inline crates/bevy_sprite/src/picking_backend.rs:125]: I think this `filter_map` is busy enough without having helper functions inline, so I would personally move that out. Or maybe it already exists somewhere in the bevy codebase?

File: crates/bevy_sprite/src/picking_backend.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+        let Some((cam_entity, camera, cam_transform, cam_ortho)) = cameras
+            .iter()
+            .filter(|(_, camera, _, _)| camera.is_active)
+            .find(|(_, camera, _, _)| {
+                camera
+                    .target
+                    .normalize(Some(match primary_window.get_single() {
+                        Ok(w) => w,
+                        Err(_) => return false,
+                    }))
```
Comment (kristoff3r) [inline crates/bevy_sprite/src/picking_backend.rs:66]: I would move the expression to a variable and use `let Ok(w) else { .. }`, it feels weird to both match and early return inside a method call.

**Review by kristoff3r — APPROVED** [review]
A few non-blocking code quality comments. I also tested the example on both linux and wasm.

File: has_sprite_ambiguities
[the remark this answers is not in the document]
Comment (jnhyatt) [reply has_sprite_ambiguities:1]: Ah shoot 😅 thanks for catching that

File: crates/bevy_sprite/src/picking_backend.rs (hunk 2 above)
Comment (jnhyatt) [reply crates/bevy_sprite/src/picking_backend.rs:125]: I looked around but couldn't find it. I originally had it at the file level, but tbh it's so generic it just looked weird. I'd be in favor of stuffing it somewhere in `bevy_math` or something.

File: crates/bevy_sprite/src/picking_backend.rs (hunk 3 above)
Comment (jnhyatt) [reply crates/bevy_sprite/src/picking_backend.rs:66]: Looking more closely, it looks like we can skip the entire function body if `primary_window.get_single()` returns `None`, can you check me on that?

File: crates/bevy_picking/src/lib.rs [hunk 4]
```diff
@@ -207,6 +207,11 @@ impl Plugin for PickingPlugin {
             .add_event::<pointer::InputPress>()
             .add_event::<pointer::InputMove>()
             .add_event::<backend::PointerHits>()
+            // Rather than try to mark all current and future backends as ambiguous with each other,
+            // we allow them to send their hits in any order. These are later sorted, so submission
+            // order doesn't matter. Other systems should *not* try to read these events in
+            // `PreUpdate` or they will silently conflict with all backends.
```
[the remark this answers is not in the document]
Comment (jnhyatt) [reply crates/bevy_picking/src/lib.rs:213]: Upon review, I think it makes sense to leave this next to the `ambiguous` call so it's not hard to find, but I think you're right that it deserves a mention in actual docs somewhere, so I think I'll split the comment between the two

File: crates/bevy_sprite/src/picking_backend.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                    blocked = is_cursor_in_sprite
+                        && pickable.map(|p| p.should_block_lower) != Some(false);
+
+                    // Transform point from world to camera space to get the Z distance
+                    let hit_pos_cam = cam_transform
+                        .affine()
+                        .inverse()
+                        .transform_point3(hit_pos_world);
+                    // HitData requires a depth as calculated from the camera's near clipping plane
+                    let depth = -cam_ortho.near - hit_pos_cam.z;
```
Comment (kristoff3r) [inline crates/bevy_sprite/src/picking_backend.rs:160]: You can also move these lets into the closure

File: crates/bevy_sprite/src/picking_backend.rs (hunk 2 above)
Comment (kristoff3r) [reply crates/bevy_sprite/src/picking_backend.rs:124]: I was thinking about where to put and thought "maybe this should actually be in glam". Well it is, here: https://docs.rs/glam/latest/glam/trait.FloatExt.html#tymethod.inverse_lerp

Comment (jnhyatt) [reply crates/bevy_sprite/src/picking_backend.rs:124]: Oh thanks! I checked glam but couldn't find it.

File: crates/bevy_sprite/src/picking_backend.rs (hunk 5 above)
Comment (jnhyatt) [reply crates/bevy_sprite/src/picking_backend.rs:160]: Ahhhhhh that's what you meant. That makes much more sense

**Pull request merged**